openmls/src/ciphersuite/mod.rs

```diff
@@ -538,6 +548,249 @@ impl Signature {
+
+            // The length of the scalar is what's left after removing the
+            // prepending zeroes, plus 1 for the msb which we've already read.
+            let mut scalar_length = scalar.len() + 1;
+
+            // If the most significant bit is 1, we have to prepend 0x00 to indicate
+            // that the integer is unsigned.
+            let prepend_zero = msb > 0x7F;
+
+            if prepend_zero {
+                // This increases the scalar length by 1.
+                scalar_length += 1;
+            };
+
+            buffer
+                // It is safe to convert to u8, because we know that the length
+                // of the scalar is at most 33.
+                .write_u8(scalar_length as u8)
+                .map_err(|_| SignatureError::EncodingError)?;
+
+            if prepend_zero {
+                // Now that we've written the length, we can actually prepend
+                // the zero.
+                buffer
+                    .write_u8(0x00)
+                    .map_err(|_| SignatureError::EncodingError)?;
+            };
+
+            // Write the msb to the encoded scalar.
+            buffer
+                .write_u8(msb)
+                .map_err(|_| SignatureError::EncodingError)?;
+
+            // Write the rest of the scalar.
+            buffer
+                .write_all(scalar)
+                .map_err(|_| SignatureError::EncodingError)?;
+
+            Ok(())
+        }
+
+        // Check overall length
+        if raw_signature.len() != 2 * P256_SCALAR_LENGTH {
+            return Err(SignatureError::EncodingError);
+        }
+
+        // We DER encode the ECDSA signature as per spec, assuming that
+        // `sign` returns two concatenated values (r||s).
+        let r = raw_signature
+            .get(..P256_SCALAR_LENGTH)
+            .ok_or(SignatureError::EncodingError)?;
+        let s = raw_signature
+            .get(P256_SCALAR_LENGTH..2 * P256_SCALAR_LENGTH)
+            .ok_or(SignatureError::EncodingError)?;
+
+        let mut encoded_signature: Vec<u8> = Vec::with_capacity(2 + 2 * (P256_SCALAR_LENGTH + 3));
+
+        // Write the DER Sequence tag
+        encoded_signature
+            .write_u8(SEQUENCE_TAG)
+            .map_err(|_| SignatureError::EncodingError)?;
+
+        // Write a placeholder byte for length. This will be overwritten once we
+        // have encoded the scalars and know the final length.
+        encoded_signature
+            .write_u8(0x00)
+            .map_err(|_| SignatureError::EncodingError)?;
+
+        encode_scalar(&r, &mut encoded_signature)?;
+        encode_scalar(&s, &mut encoded_signature)?;
+
+        // Now we can determine the length of the scalars by taking the overall
+        // length and removing 1 for the Sequence tag and 1 for the length
+        // field. The conversion to u8 is safe, because we know that each of the
+        // scalars is at most 33 bytes long.
+        *encoded_signature
+            .get_mut(1)
```

Comment:
Ah I see how you do this now. Another option to do this is to first get the length and write it directly. I suspect that this would be faster (because you can allocate the memory you need at once). But I'm fine with leaving it like this for now.